R/TealAppDriver.R

```diff
@@ -554,18 +553,77 @@ TealAppDriver <- R6::R6Class( # nolint: object_name.
       filter_panel = character(0)
     ),
     # private methods ----
-    set_active_ns = function() {
+    # Helper function to extract wrapper ID from selector and take first match if multiple found
+    extract_wrapper_id = function(selector) {
+      wrapper_id <- sub(
+        "^#",
+        "",
+        self$get_attr(selector = selector, attribute = "href")
+      )
+      # Take first match if multiple found
+      if (length(wrapper_id) > 1) {
+        wrapper_id <- wrapper_id[1]
+      }
+      wrapper_id
+    },
+    # Helper function to check if wrapper ID is valid
+    is_valid_wrapper_id = function(wrapper_id) {
+      length(wrapper_id) == 1 && wrapper_id != "" && !is.na(wrapper_id)
+    },
+    set_active_ns = function(sleep_time = 0.5) {
+      # Although wait_for_idle() is called before set_active_ns(), it only ensures Shiny is not processing.
+      # wait_for_page_stability() is needed here to ensure the DOM/UI is fully rendered and stable
+      # before trying to extract the namespace.
+      private$wait_for_page_stability()
+
       all_inputs <- self$get_values()$input
       active_tab_inputs <- all_inputs[grepl("-active_module_id$", names(all_inputs))]
 
-      active_wrapper_id <- sub(
-        "^#",
-        "",
-        self$get_attr(
-          selector = sprintf(".teal-modules-tree li a.module-button[data-value='%s']", active_tab_inputs),
-          attribute = "href"
+      # If no active_module_id input found, find the selected/active tab button directly
```

Comment:
I don't get why would DOM have the input id if the input does not exist in shiny.

How would that work?

- Is it because of `bslib` doing some magic? This creates additional overhead when/if the navigation selectors are changed.